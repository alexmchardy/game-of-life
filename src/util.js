// Utility functions

/**
 * @typedef {number[][]} CoordList - Array of coordinates (e.g [[0,0], [1,2]])
 */

// Simple cache for the most recently generated empty grid
const cachedEmpty = {
	height: null,
	width: null,
	grid: [],
};

/**
 * Return new empty grid of specified height and width
 * @param {number} h - Desired height of grid
 * @param {number} w - Desired width of grid
 * @returns {boolean[][]} Empty h x w grid
 */
function getEmptyGrid(h, w) {
	if (cachedEmpty.height !== h || cachedEmpty.width !== w) {
		cachedEmpty.height = h;
		cachedEmpty.width = w;
		cachedEmpty.grid = Array(h)
			.fill(null)
			.map(() => Array(w).fill(false));
	}
	return cachedEmpty.grid.map((row) => row.slice());
}

/**
 * Return new grid of specified size with live cells
 * @param {Object} [opts={}]
 * @param {number} [opts.size=5] - Desired size (height and width) of grid
 * @param {number} [opts.height=5] - Desired height of grid
 * @param {number} [opts.width=5] - Desired width of grid
 * @param {CoordList} [opts.live=[]] - Array of coordinates of live cells
 * @returns {boolean[][]} Grid of h x w size with live cells populated
 */
function grid({ size = 5, height, width, live = [] } = {}) {
	const h = height || size;
	const w = width || size;
	const newGrid = getEmptyGrid(h, w);

	// Set live cells that are within the grid dimensions
	live.forEach(([x, y]) => {
		if (x < w && y < h) {
			newGrid[x][y] = true;
		}
	});

	return newGrid;
}

export { getEmptyGrid, grid };
