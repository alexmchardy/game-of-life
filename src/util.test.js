import { grid, getEmptyGrid } from './util';

describe('grid', () => {
	test('returns default grid', () => {
		const g = grid();
		expect(g).toMatchSnapshot();
	});

	test('returns grid with one specified dimension', () => {
		const g = grid({ size: 3 });
		expect(g).toHaveLength(3);
		expect(g[0]).toHaveLength(3);
	});

	test('returns g with both specified dimensions', () => {
		const g = grid({ height: 3, width: 2 });
		expect(g).toHaveLength(3);
		expect(g[0]).toHaveLength(2);
	});

	test('returns grid initialized with live cells', () => {
		const g = grid({
			live: [
				[2, 2],
				[3, 4],
			],
		});
		expect(g[2][2]).toEqual(true);
		expect(g[3][4]).toEqual(true);
	});
});

describe('getEmptyGrid', () => {
	test('returns empty grid', () => {
		const g = getEmptyGrid(4, 5);
		expect(g).toHaveLength(4);
		expect(g[0]).toHaveLength(5);
		expect(g[0][0]).toEqual(false);
	});

	test('caches and reuses empty grid when called twice with same dimensions', () => {
		const fillSpy = jest.spyOn(Array.prototype, 'fill');
		const g1 = getEmptyGrid(2, 3);
		expect(fillSpy).toHaveBeenCalledTimes(3);
		fillSpy.mockClear();
		const g2 = getEmptyGrid(2, 3);
		// Cached grid was used so Array.fill() should not be called again
		expect(fillSpy).toHaveBeenCalledTimes(0);
		expect(g2).toHaveLength(2);
		expect(g2[0]).toHaveLength(3);
		// Make sure cached grid was cloned (not a ref to same array)
		expect(g2).not.toBe(g1);
	});
});
