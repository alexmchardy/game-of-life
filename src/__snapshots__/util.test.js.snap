// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`grid returns default grid 1`] = `
Array [
  Array [
    false,
    false,
    false,
    false,
    false,
  ],
  Array [
    false,
    false,
    false,
    false,
    false,
  ],
  Array [
    false,
    false,
    false,
    false,
    false,
  ],
  Array [
    false,
    false,
    false,
    false,
    false,
  ],
  Array [
    false,
    false,
    false,
    false,
    false,
  ],
]
`;
